File: backend/app/auth/pow.py
"""Proof-of-Work challenge for registration anti-bot protection.

Client must find a nonce such that SHA256(challenge + nonce) starts with N hex zeros.
Difficulty 4 ≈ 65k iterations ≈ 50-200ms in browser, blocks simple bots.
"""

import hashlib
import secrets
import time

_challenges: dict[str, float] = {}  # challenge → created_at (monotonic)
DIFFICULTY = 4  # number of leading hex zeros required
CHALLENGE_TTL = 300  # 5 minutes


def create_challenge() -> str:
    """Generate a new PoW challenge."""
    # Cleanup expired challenges
    now = time.monotonic()
    expired = [k for k, v in _challenges.items() if now - v > CHALLENGE_TTL]
    for k in expired:
        _challenges.pop(k, None)

    challenge = secrets.token_hex(16)
    _challenges[challenge] = now
    return challenge


def verify_pow(challenge: str, nonce: str) -> bool:
    """Verify a PoW solution. Each challenge can only be used once."""
    if not challenge or not nonce:
        return False
    ts = _challenges.pop(challenge, None)
    if ts is None:
        return False
    if time.monotonic() - ts > CHALLENGE_TTL:
        return False
    h = hashlib.sha256(f"{challenge}{nonce}".encode()).hexdigest()
    return h[:DIFFICULTY] == "0" * DIFFICULTY

File: frontend-fiction/src/lib/pow.ts
/**
 * Proof-of-Work solver for registration anti-bot protection.
 * Finds a nonce such that SHA256(challenge + nonce) starts with `difficulty` hex zeros.
 * ~50-200ms in browser for difficulty=4.
 */
export async function solveChallenge(challenge: string, difficulty = 4): Promise<string> {
  const target = '0'.repeat(difficulty);
  for (let nonce = 0; ; nonce++) {
    const data = new TextEncoder().encode(challenge + nonce);
    const hash = await crypto.subtle.digest('SHA-256', data);
    const hex = [...new Uint8Array(hash)]
      .map((b) => b.toString(16).padStart(2, '0'))
      .join('');
    if (hex.startsWith(target)) return String(nonce);
  }
}
